### PR TITLE
Add x-originalName handling in spec and code generator

### PR DESCRIPTION
**Pull request by RicoSuter**
If the C# controller parameter name does not match the HTTP parameter name then generate a custom name so that the client gens can generate the same parameter names.

### CODE REVIEW

File: src/NSwag.Core/OpenApiParameter.cs [hunk 1]
```diff
@@ -44,6 +44,10 @@ public string Name
             }
         }
 
+        /// <summary>Gets or sets a custom name which is often used in code generation (default: null).</summary>
+        [JsonProperty(PropertyName = "x-name", DefaultValueHandling = DefaultValueHandling.IgnoreAndPopulate)]
```
Comment (RicoSuter) [inline src/NSwag.Core/OpenApiParameter.cs:48]: Better use x-parameterName to avoid collisions with other tools.

Comment (jeremyVignelles) [reply src/NSwag.Core/OpenApiParameter.cs:48]: `x-parameter-name` or `x-original-parameter` would be better names IMHO

File: src/NSwag.Generation.AspNetCore/Processors/OperationParameterProcessor.cs [hunk 2]
```diff
@@ -184,8 +184,8 @@ public bool Process(OperationProcessorContext operationProcessorContext)
 
                 if (operationParameter != null)
                 {
-                    operationParameter.Position = position;
                     position++;
+                    operationParameter.Position = position;
```
Comment (jeremyVignelles) [inline src/NSwag.Generation.AspNetCore/Processors/OperationParameterProcessor.cs:188]: What's the rationale for this change?

File: src/NSwag.Generation.WebApi/Processors/OperationParameterProcessor.cs [hunk 3]
```diff
@@ -182,8 +182,18 @@ public bool Process(OperationProcessorContext context)
 
                 if (operationParameter != null)
                 {
-                    operationParameter.Position = position;
                     position++;
+                    operationParameter.Position = position;
+
+                    if (_settings.SchemaType == SchemaType.OpenApi3)
+                    {
+                        operationParameter.IsNullableRaw = null;
```
Comment (jeremyVignelles) [inline src/NSwag.Generation.WebApi/Processors/OperationParameterProcessor.cs:190]: What does this mean ?

File: src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs [hunk 4]
```diff
@@ -21,19 +21,22 @@ public class DefaultParameterNameGenerator : IParameterNameGenerator
         /// <returns>The parameter name.</returns>
         public string Generate(OpenApiParameter parameter, IEnumerable<OpenApiParameter> allParameters)
         {
-            if (string.IsNullOrEmpty(parameter.Name))
+            var name = !string.IsNullOrEmpty(parameter.CustomName) ?
```
[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs:24]: Didn't understand the question...

[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs:24]: I get it now. I don't think it is necessary if the spec parameter is named anything else than `x-name` : it is explicit that it is generated by NSwag and I'd argue that the parameter should match on both side.

What's the breaking risk of that change?

File: src/NSwag.Generation.WebApi/Processors/OperationParameterProcessor.cs (hunk 3 above)
Comment (RicoSuter) [reply src/NSwag.Generation.WebApi/Processors/OperationParameterProcessor.cs:190]: This is a fix to align web api impl with asp.net core - not really part of this PR.

File: src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (RicoSuter) [reply src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs:24]: > What's the breaking risk of that change?

Not really a break, in the current PR state you can enable/disable in the spec gen but not in the client gen - just wondering whether we should have a config for both so it's aligned.

File: src/NSwag.Core/OpenApiParameter.cs (hunk 1 above)
Comment (RicoSuter) [reply src/NSwag.Core/OpenApiParameter.cs:48]: All properties in JSON Schema/OpenAPI use lower camel case, e.g. "securityDefinitions" - so i would go with x-lowerCamelCase, no?

I like x-originalParameter better.

File: src/NSwag.Generation.AspNetCore/Processors/OperationParameterProcessor.cs (hunk 2 above)
Comment (RicoSuter) [reply src/NSwag.Generation.AspNetCore/Processors/OperationParameterProcessor.cs:188]: Nothing, looks better to me :-)

File: src/NSwag.Core/OpenApiParameter.cs (hunk 1 above)
Comment (jeremyVignelles) [reply src/NSwag.Core/OpenApiParameter.cs:48]: Then ok for `x-originalParameter`

File: src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs:24]: In the current release, what happens if I try to create an API with a different name ? Which name is taken, the param name or the HTTP one ?

The reason why we would introduce another parameter would be to make the new behavior opt-it to not break the current implementations.

File: src/NSwag.Generation.AspNetCore/Processors/OperationParameterProcessor.cs (hunk 2 above)
Comment (jeremyVignelles) [reply src/NSwag.Generation.AspNetCore/Processors/OperationParameterProcessor.cs:188]: Is Position supposed to be 0-based or 1-based ?

Comment (jeremyVignelles) [reply src/NSwag.Generation.AspNetCore/Processors/OperationParameterProcessor.cs:188]: with `position = 1` above, that will make you start at 2, which I'm not sure is the wanted start number :) Did you mean to set `position = 0` instead ?

Comment (RicoSuter) [reply src/NSwag.Generation.AspNetCore/Processors/OperationParameterProcessor.cs]: Oh shit, yes, reverted.

File: src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (RicoSuter) [reply src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs:24]: > Which name is taken, the param name or the HTTP one ?

Currently it always uses the HTTP name and the other is ignored (ie from an API perspective it doesnt matter). Currently there is no way to preserve the original parameter name.

[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply src/NSwag.CodeGeneration/DefaultParameterNameGenerator.cs]: If I have http parameter A and C# parameter B, the name A was used in generated client code, and now it's named B.
It might be breaking in cases like these:
```cs
await client.CallMethod(A: 123);
```

I don't know if it's really a big deal though, that kind of cases should be very rare.

Do you see any other potential breaking change ?
